src/controller/CHIPDevice.h

```diff
@@ -0,0 +1,187 @@
+ *    messages to and from the corresponding CHIP devices.
+ */
+
+#pragma once
+
+#include <core/CHIPCore.h>
+#include <support/Base64.h>
+#include <support/DLLUtil.h>
+#include <transport/SecurePairingSession.h>
+#include <transport/SecureSessionMgr.h>
+#include <transport/raw/MessageHeader.h>
+#include <transport/raw/UDP.h>
+
+namespace chip {
+namespace Controller {
+
+class DeviceController;
+class DeviceStatusDelegate;
+struct SerializedDevice;
+
+class DLL_EXPORT Device
+{
+public:
+    Device() {}
+    ~Device() {}
+
+    void SetDelegate(DeviceStatusDelegate * delegate) { mStatusDelegate = delegate; }
+
+    // ----- Messaging -----
+    /**
+     * @brief
+     *   Send the provided message to the device
+     *
+     * @param[in] message   The message to be sent. The ownership of the message buffer
+     *                      is handed over to Device object. SendMessage() will
+     *                      decrement the reference count of the message buffer before
+     *                      returning.
+     *
+     * @return CHIP_ERROR   CHIP_NO_ERROR on success, or corresponding error
+     */
+    CHIP_ERROR SendMessage(System::PacketBuffer * message);
+
+    /**
+     * @brief
+     *   Get the IP address assigned to the device.
+     *
+     * @param[out] addr   The reference to the IP address.
+     *
+     * @return true, if the IP address was filled in the out parameter, false otherwise
+     */
+    bool GetIpAddress(Inet::IPAddress & addr) const;
+
+    void Init(SecureSessionMgr<Transport::UDP> * sessionMgr, Inet::InetLayer * inetLayer)
+    {
+        mSessionManager = sessionMgr;
+        mInetLayer      = inetLayer;
+    }
+
+    void Init(SecureSessionMgr<Transport::UDP> * sessionMgr, Inet::InetLayer * inetLayer, NodeId deviceId, uint16_t devicePort,
+              Inet::InterfaceId interfaceId)
+    {
+        Init(sessionMgr, inetLayer);
+        mDeviceId   = deviceId;
+        mDevicePort = devicePort;
+        mInterface  = interfaceId;
+        mState      = kConnectionState_Connecting;
+    }
+
+    /** @brief Serialize the Pairing Session to a string.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR Serialize(SerializedDevice & output);
+
+    /** @brief Deserialize the Pairing Session from the string.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR Deserialize(const SerializedDevice & input);
+
+    void OnMessageReceived(const PacketHeader & header, const PayloadHeader & payloadHeader, Transport::PeerConnectionState * state,
+                           System::PacketBuffer * msgBuf, SecureSessionMgrBase * mgr);
+
+    bool IsActive() { return mActive; }
+
+    void SetActive(bool active) { mActive = active; }
+
+    NodeId GetDeviceId() { return mDeviceId; }
+
+    void SetAddress(Inet::IPAddress deviceAddr) { mDeviceAddr = deviceAddr; }
```

Comment:
I think IPAddress may be larger than a basic data type. Better to pass it as const&